### PR TITLE
CEP-187 Added implementation for tenant-metrics API

### PR DESCRIPTION
**Pull request by arpitamishra-macrometa**
This pull request contains the implementation for extending c84j library with tenant-metrics API (URL: {base-url}/_fabric/{fabric}/_api/metrics/query.

### CODE REVIEW

File: src/main/java/com/c8db/internal/InternalC8Admin.java [hunk 1]
```diff
@@ -95,5 +110,11 @@ protected Request getTenantFeaturesRequest(final String tenant) {
         return request(db.tenant(), db.name(), RequestType.GET, PATH_API_FEATURES, PATH_TENANT, tenant);
     }
 
-
+    protected Request getTenantMetricsRequest(final String tenant, TenantMetricsOption options){
+        //Finalize defaults of TenantMetricsOption with Stoyan
```
Comment (grainier-macrometa) [inline src/main/java/com/c8db/internal/InternalC8Admin.java:114]: Let's make this a TODO, by adding `// TODO: `at the beginning.

File: src/main/java/com/c8db/internal/InternalC8Admin.java [hunk 2]
```diff
@@ -95,5 +110,11 @@ protected Request getTenantFeaturesRequest(final String tenant) {
         return request(db.tenant(), db.name(), RequestType.GET, PATH_API_FEATURES, PATH_TENANT, tenant);
     }
 
-
+    protected Request getTenantMetricsRequest(final String tenant, TenantMetricsOption options){
```
Comment (grainier-macrometa) [inline src/main/java/com/c8db/internal/InternalC8Admin.java:113]: `final String tenant` Seems redundant.

**Review by grainier-macrometa — CHANGES_REQUESTED** [review]

File: src/main/java/com/c8db/entity/MetricsEntity.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package com.c8db.entity;
+
```
Comment (StoyanTc) [inline src/main/java/com/c8db/entity/MetricsEntity.java:2]: Can you please make this class nested for TenantMetricsEntity? It won't be used anywhere else.

File: src/main/java/com/c8db/internal/C8AdminImpl.java [hunk 4]
```diff
@@ -39,4 +45,12 @@ public LimitsEntity getTenantLimits(String tenant) throws C8DBException {
 	public TenantEntity getTenant(String tenant) throws C8DBException {
 		return executor.execute(getTenantRequest(tenant), getTenantResponseDeserializer());
 	}
+
+	@Override
+	public TenantMetricsEntity getTenantMetrics(String tenant, TenantMetricsOption options) throws C8DBException, JsonProcessingException {
+		String result =  executor.execute(getTenantMetricsRequest(tenant,options),
+				getTenantMetricResponseDeserializer());
+		TenantMetricsEntity tenantMetrics = C8Mapper.mapTenantMetrics(result);
```
Comment (StoyanTc) [inline src/main/java/com/c8db/internal/C8AdminImpl.java:53]: Please change the `getTenantMetricResponseDeserializer()` to return straight an object of `TenantMetricsEntity` and get rid of `C8mapper`?

You could refer to `getTenantLimitsResponseDeserializer()` implementation.

File: src/main/java/com/c8db/internal/InternalC8Admin.java (hunk 1 above)
Comment (StoyanTc) [reply src/main/java/com/c8db/internal/InternalC8Admin.java:114]: And you might avoid my name into the comment as well? :)

**Review by StoyanTc — CHANGES_REQUESTED** [review]

File: src/main/java/com/c8db/C8Admin.java [hunk 5]
```diff
@@ -4,11 +4,10 @@
 
 package com.c8db;
 
-import com.c8db.entity.ApiKeyEntity;
-import com.c8db.entity.FeaturesEntity;
-import com.c8db.entity.LimitsEntity;
-import com.c8db.entity.TenantEntity;
-import com.c8db.entity.TenantsEntity;
+import com.c8db.entity.*;
```
[the remark this answers is not in the document]
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/C8Admin.java:7]: Added single imports as suggested. Changes will be pushed with next commit.

File: src/main/java/com/c8db/entity/MetricsEntity.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package com.c8db.entity;
```
[the remark this answers is not in the document]
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/entity/MetricsEntity.java:1]: Added copyright information in the newly added classes. Changes will be pushed with next commit.

File: src/main/java/com/c8db/C8Admin.java [hunk 7]
```diff
@@ -59,6 +58,15 @@
...
     LimitsEntity getTenantLimits(final String tenant) throws C8DBException;
 
+    /**
+     * Returns metrics for a tenant
+     *  @param tenant The tenant name
+     *  @param options The parameters passed as a part of request body
+     *  @return metrics for this tenant
+     *  @throws C8DBException
+     */
+    TenantMetricsEntity getTenantMetrics(final String tenant, TenantMetricsOption options) throws C8DBException, JsonProcessingException;
```
[the remark this answers is not in the document]
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/C8Admin.java:68]: Implemented.

File: src/main/java/com/c8db/internal/C8AdminImpl.java [hunk 8]
```diff
@@ -6,11 +6,17 @@
 
 import com.c8db.C8Admin;
 import com.c8db.C8DBException;
-import com.c8db.entity.FeaturesEntity;
-import com.c8db.entity.LimitsEntity;
-import com.c8db.entity.TenantEntity;
-import com.c8db.entity.TenantsEntity;
-
+import com.c8db.entity.*;
```
[the remark this answers is not in the document]
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/C8AdminImpl.java:9]: Implemented.

File: src/main/java/com/c8db/internal/InternalC8Admin.java [hunk 9]
```diff
@@ -70,15 +72,28 @@ public FeaturesEntity deserialize(final Response response) throws VPackException
     }
 
     protected ResponseDeserializer<LimitsEntity> getTenantLimitsResponseDeserializer() {
+        System.out.println("Building response....");
```
[the remark this answers is not in the document]
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/InternalC8Admin.java:75]: Removed println.

File: src/main/java/com/c8db/internal/InternalC8Admin.java (hunk 1 above)
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/InternalC8Admin.java:114]: Implemented.

File: src/main/java/com/c8db/internal/InternalC8Admin.java (hunk 2 above)
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/InternalC8Admin.java:113]: Resolved.

File: src/main/java/com/c8db/entity/MetricsEntity.java (hunk 3 above)
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/entity/MetricsEntity.java:2]: Implemented.

File: src/main/java/com/c8db/internal/C8AdminImpl.java (hunk 4 above)
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/C8AdminImpl.java]: Implemented.

File: src/main/java/com/c8db/internal/InternalC8Admin.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (grainier-macrometa) [reply src/main/java/com/c8db/internal/InternalC8Admin.java:85]: `println` is still there, right?

[the remark this answers is not in the document]
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/InternalC8Admin.java]: Fixed.

File: src/main/java/com/c8db/C8Admin.java [hunk 10]
```diff
@@ -4,11 +4,15 @@
...
 
-import com.c8db.entity.ApiKeyEntity;
+
 import com.c8db.entity.FeaturesEntity;
 import com.c8db.entity.LimitsEntity;
 import com.c8db.entity.TenantEntity;
 import com.c8db.entity.TenantsEntity;
+import com.c8db.entity.TenantMetricsEntity;
+import com.c8db.model.TenantMetricsOption;
+import com.fasterxml.jackson.core.JsonProcessingException;
```
Comment (StoyanTc) [inline src/main/java/com/c8db/C8Admin.java:14]: @arpitamishra-macrometa - Please get rid of unused import.

File: src/main/java/com/c8db/internal/InternalC8Admin.java [hunk 11]
```diff
@@ -71,10 +83,23 @@ public FeaturesEntity deserialize(final Response response) throws VPackException
...
+    }
+
+    protected ResponseDeserializer<TenantMetricsEntity> getTenantMetricResponseDeserializer() {
+        return new ResponseDeserializer<TenantMetricsEntity>() {
+            @Override
+            public TenantMetricsEntity deserialize(final Response response) throws VPackException {
+                final VPackSlice result = response.getBody().get(C8ResponseField.RESULT);
+                String responseObj = util().deserialize(result,  new Type<String>(){}.getType());
+                TenantMetricsEntity tenantMetrics = getTenantMetrics(responseObj);
+                return tenantMetrics;
```
Comment (StoyanTc) [inline src/main/java/com/c8db/internal/InternalC8Admin.java:102]: @arpitamishra-macrometa - Won't it work here the following construct:
```
TenantMetricsEntity tenantMetrics =  util().deserialize(result,  new Type<TenantMetricsEntity >(){}.getType());```

Do you really need to map the response twice - first to `String` and then to `TenantMetricsEntity `?

Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/InternalC8Admin.java:102]: I tried this way earlier , but this won't work as the response contains the value against key "c8cep_app_throughput_total" and "c8cep_app_latency_seconds_sum" nested inside "results" . Hence I have added this extra processing logic.

Comment (StoyanTc) [reply src/main/java/com/c8db/internal/InternalC8Admin.java:102]: ok, @arpitamishra-macrometa I'll give a try too a bit later.

File: src/main/java/com/c8db/C8Admin.java (hunk 10 above)
Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/C8Admin.java]: Removed used import for JsonProcessingException

File: src/main/java/com/c8db/internal/InternalC8Admin.java (hunk 11 above)
Comment (StoyanTc) [reply src/main/java/com/c8db/internal/InternalC8Admin.java:102]: @arpitamishra-macrometa - Can you please try once again using annotation of the real fields' names, e.g.
```
public class TenantMetricsEntity implements Entity {
        @SerializedName("c8cep_app_throughput_total")
        List<MetricsEntity> throughput;
        @SerializedName("c8cep_app_latency_seconds_sum")
        List<MetricsEntity> latencySum;
        ....
```

Comment (arpitamishra-macrometa) [reply src/main/java/com/c8db/internal/InternalC8Admin.java:102]: This implementation didn't work as well.

**Review by StoyanTc — APPROVED** [review]

**Review by grainier-macrometa — APPROVED** [review]

**Pull request merged**